google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -27,49 +29,113 @@ module Spanner
       # {Google::Cloud::Spanner::Session} instances.
       #
       class Pool
-        attr_accessor :min, :max, :pool, :queue
+        attr_accessor :all_sessions, :session_queue, :transaction_queue
 
-        def initialize client, min: 2, max: 10, keepalive: 1500
+        def initialize client, min: 2, max: 10, keepalive: 1500,
+                       write_ratio: 0.5, fail: true
           @client = client
           @min = min
           @max = max
           @keepalive = keepalive
-          @pool = []
-          @queue = []
+          @write_ratio = write_ratio
+          @write_ratio = 0 if write_ratio < 0
+          @write_ratio = 1 if write_ratio > 1
+          @fail = fail
+
+          @mutex = Mutex.new
+          @resource = ConditionVariable.new
 
           # initialize pool and availability queue
           init
         end
 
         def with_session
-          session = checkout
-          yield session
-          checkin session
+          session = checkout_session
+          begin
+            yield session
+          ensure
+            checkin_session session
+          end
+        end
+
+        def checkout_session
+          @mutex.synchronize do
+            loop do
+              read_session = session_queue.shift
+              return read_session if read_session
+              write_transaction = transaction_queue.shift
+              return write_transaction.session if write_transaction
+
+              return new_session! if can_allocate_more_sessions?
+
+              fail SessionLimitError if @fail
+
+              @resource.wait @mutex
+            end
+          end
+        end
+
+        def checkin_session session
+          unless all_sessions.include? session
+            fail ArgumentError, "Cannot checkin session"
+          end
+
+          @mutex.synchronize do
+            session_queue.push session
+
+            @resource.broadcast
+          end
+
+          nil
         end
 
-        def checkout
-          if queue.empty?
-            fail "No available sessions" if pool.size >= @max
-            new_session!
+        def with_transaction
+          tx = checkout_transaction
+          begin
+            yield tx
+          ensure
+            checkin_transaction tx
           end
-          queue.shift
         end
 
-        def checkin session
-          fail "Cannot checkin session" unless pool.include? session
+        def checkout_transaction
+          @mutex.synchronize do
+            loop do
+              write_transaction = transaction_queue.shift
+              return write_transaction if write_transaction
+              read_session = session_queue.shift
+              return read_session.create_transaction if read_session
```

Comment:
I guess create_transaction makes an rpc? In that case you should not hold mutex for all that time.